File: android/wallpaperinfo/app/src/main/java/com/thirdwavesoft/wallpaperinfo/PlatformTimer.java
package com.thirdwavesoft.wallpaperinfo;

import android.util.Log;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;


public class PlatformTimer {
	
	public PlatformTimer(WallpaperService service) {
		_wallpaperService = service;
	}

    synchronized private void scheduleSwitchWallpaper(int seconds) {
   		if (_switchHandler != null) {
   			cancelSwitchWallpaper();
   		}
   		final Runnable wallpaperSwitcher = new Runnable() {
   			public void run() {
			// Log.d(_TAG, "wallpaperSwitcher()");
			synchronized(this) {
				try {
					if (!_pause) {
						_wallpaperService.wallpaperSwitchCallback();
					}
				} catch (Exception e) {
					e.printStackTrace();
					Log.e(_TAG, e.getMessage());
				}
			}
   			}
   		};
   		_pause = false;
   		_switchHandler = _scheduler.scheduleWithFixedDelay(wallpaperSwitcher, seconds, seconds, TimeUnit.SECONDS);
   	}
   	
    synchronized public void cancelSwitchWallpaper() {
    	if (_switchHandler != null) {
	   		_switchHandler.cancel(true);
	   		_switchHandler = null;
    	}
   		_pause = true;
   	}
    
   	synchronized public void resetTimer() {
   		if (!_pause) {
   			resume();
   		}
   	}
   	
   	synchronized public void pause() {
		cancelSwitchWallpaper();
   	}
   	
   	synchronized public void resume() {
   		scheduleSwitchWallpaper(_interval);
   	}
   	
   	synchronized public void setInterval(int interval) {
   		_interval = interval;
   		if (!_pause) {
   			resume();
   		}
   	}
   	
   	private WallpaperService _wallpaperService;
	private ScheduledFuture<?> _switchHandler = null;
	private int _interval = 5;
	private boolean _pause = false;
	private final ScheduledExecutorService _scheduler =	Executors.newScheduledThreadPool(1);

	private static final String _TAG = "PlatformTimer";

}

File: ios_tvos/tromso/tromso/PlatformTimer.swift
//
//  WallpaperSwitchScheduler.swift
//  WallpaperInfoApp
//
//  Created by Brad Park on 4/11/20.
//  Copyright © 2020 Brad Park. All rights reserved.
//

import Foundation

class PlatformTimer {

    init(_ wallpaperService: WallpaperService) {
        _wallpaperService = wallpaperService
    }

    private func scheduleSwitchWallpaper(seconds: Int) {
        if (_switchHandler != nil) {
            cancelSwitchWallpaper();
        }
        _switchHandler = RepeatingTimer(timeInterval: TimeInterval(seconds))
        _switchHandler!.eventHandler = {
			print("PlatformTimer.eventHandler{}")
			if (!self._pause) {
				self._wallpaperService.wallpaperSwitchCallback();
			}
        }
        _pause = false;
        _switchHandler!.resume()
    }
    
    func cancelSwitchWallpaper() {
        if (_switchHandler != nil) {
			_switchHandler?.suspend()
            _switchHandler = nil;
        }
        _pause = true
    }
    
    func resetTimer() {
        if (!_pause) {
            resume();
        }
    }
    
    func pause() {
		// print("PlatformTimer.pause()")
        cancelSwitchWallpaper();
    }
    
    func resume() {
		// print("PlatformTimer.resume()")
        scheduleSwitchWallpaper(seconds: _interval);
    }
    
    func setInterval(interval: Int) {
        _interval = interval;
        if (!_pause) {
            resume();
        }
    }
    
    private var _wallpaperService: WallpaperService;
    private var _switchHandler:RepeatingTimer? = nil;
    private var _interval:Int = 5;
    private var _pause:Bool = false;

}
